clients/src/test/java/org/apache/kafka/common/network/TransportLayerUtils.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.network;
+
+import org.apache.kafka.common.errors.AuthenticationException;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.SocketChannel;
+import java.security.Principal;
+
+public final class TransportLayerUtils {
+    private TransportLayerUtils() {
+    }
+
+    /**
+     * Create a TransportLayer based on GatheringByteChannel. This helper method is used by testing only and noted that
+     * not all methods of TransportLayer are supported.
+     * @param channel channel
+     * @return a TransportLayer
+     */
+    public static TransportLayer toTransportLayer(GatheringByteChannel channel) {
```

Comment:
I had tried making this change, but it turns out we still have places where we use `GatheringByteChannel`, so not sure it's worth it.

Reply:
The benefit, personally, is that the ugly casting is removed from production code. However, you are right - ```GatheringByteChannel``` still makes us leave a wrapper in test scope...

Reply:
If we move the casting in `FileRecords` to `TransportLayers`, then it would be isolated in one class. Also, it's safe casting, not unsafe. That may be OK?

Reply:
```java
        if (destChannel instanceof TransportLayer) {
            TransportLayer tl = (TransportLayer) destChannel;
            bytesTransferred = tl.transferFrom(channel, position, count);
        } else {
            bytesTransferred = channel.transferTo(position, count, destChannel);
        }
```
Did you mean that we could move above code to ```TransportLayers```. If so, it seems that we still don't replace ```GatheringByteChannel``` by ```TransportLayer``` in production code.

Most usage of ```GatheringByteChannel``` implementation (in test scope) is ```ByteBufferChannel```. We can rewrite ```ByteBufferChannel``` to ```ByteBufferTransportLayer``` so the ```toTransportLayer``` wrapper can be removed.
Also, the new ```ByteBufferTransportLayer ``` should be moved to test code as it is not used by production code.

Reply:
@ijuma please take a look at https://github.com/apache/kafka/pull/9516/commits/9369c7a67a66f55e1f0a93d8414d03a6e10ea410 to see what I described if you have free cycle :)